components/layout/flex.rs

```diff
@@ -0,0 +1,445 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Layout for elements with a CSS `display` property of `flex`.
+
+#![deny(unsafe_code)]
+
+use block::BlockFlow;
+use context::LayoutContext;
+use display_list_builder::FlexFlowDisplayListBuilding;
+use floats::FloatKind;
+use flow;
+use flow::{Flow, FlowClass, OpaqueFlow};
+use flow::{HAS_LEFT_FLOATED_DESCENDANTS, HAS_RIGHT_FLOATED_DESCENDANTS};
+use flow::ImmutableFlowUtils;
+use flow::INLINE_POSITION_IS_STATIC;
+use flow::IS_ABSOLUTELY_POSITIONED;
+use flow::mut_base;
+use fragment::{Fragment, FragmentBorderBoxIterator};
+use incremental::{REFLOW, REFLOW_OUT_OF_FLOW};
+use layout_debug;
+use model::{IntrinsicISizes};
+use style::computed_values::{flex_direction, float};
+use style::properties::style_structs;
+use style::values::computed::LengthOrPercentageOrAuto;
+
+use euclid::{Point2D, Rect};
+use gfx::display_list::DisplayList;
+use std::cmp::max;
+use std::sync::Arc;
+use style::properties::ComputedValues;
+use util::geometry::Au;
+use util::logical_geometry::LogicalSize;
+use util::opts;
+use model::MaybeAuto;
+
+// A mode describes which logical axis a flex axis is parallel with.
+// The logical axises are inline and block, the flex axises are main and cross.
+// When the flex container has flex-direction: column or flex-direction: column-reverse, the main axis
+// should be block. Otherwise, it should be inline.
+#[derive(Debug)]
+enum Mode {
+    Inline,
+    Block
+}
+
+/// A block with the CSS `display` property equal to `flex`.
+#[derive(Debug)]
+pub struct FlexFlow {
+    /// Data common to all block flows.
+    block_flow: BlockFlow,
+    /// The logical axis which the main axis will be parallel with.
+    /// The cross axis will be parallel with the opposite logical axis.
+    main_mode: Mode,
+}
+
+fn flex_style(fragment: &Fragment) -> &style_structs::Flex {
+    fragment.style.get_flex()
+}
+
+// TODO(zentner): This function should use flex-basis.
+fn flex_item_inline_sizes(flow: &mut Flow) -> IntrinsicISizes {
+    let _scope = layout_debug_scope!("flex::flex_item_inline_sizes");
+    debug!("flex_item_inline_sizes");
+    let base = flow::mut_base(flow);
+
+    debug!("FlexItem intrinsic inline sizes: {:?}, {:?}",
+           base.intrinsic_inline_sizes.minimum_inline_size,
+           base.intrinsic_inline_sizes.preferred_inline_size);
+
+    IntrinsicISizes {
+        minimum_inline_size: base.intrinsic_inline_sizes.minimum_inline_size,
+        preferred_inline_size: base.intrinsic_inline_sizes.preferred_inline_size,
+    }
+}
+
+impl FlexFlow {
+    pub fn from_fragment(fragment: Fragment,
+                         flotation: Option<FloatKind>)
+                         -> FlexFlow {
+
+        let main_mode = match flex_style(&fragment).flex_direction {
+            flex_direction::T::row_reverse    => Mode::Inline,
+            flex_direction::T::row            => Mode::Inline,
```

Comment:
nit: you can use or-patterns (`flex_direction::T::row_reverse | flex_direction::T::row => ...`) to avoid a little bit of code duplication here